### DISCUSSION TITLE
DOM interaction questions

### DISCUSSION THREAD

**Discussion by ghost** [pyodide/pyodide#4733] [Q&A]
Very often on the first (one, two, three...) clicks on a DOM element, the event handler doesn't see node attributes that are definitely there (confirmed from the browser's dev tools), and therefore returns None/null when these attributes are queried from Python pyodide code. Some clicks later, or after clicking on another element it works again.

I have the strong feeling that the WASM Code runs in a blocking thread and doesn't see updates to the DOM, and I cannot find a programmatic way to make my code see the updated DOM while it is running.

The same problem applies when I try to dynamically create DOM nodes, and from these DOM nodes get callbacks into my code. It just doesn't seem possible.

My workaround is creating an IFRAME element before leaving the DOM gen code, which after display loads another Python callback that can then react, create more DOM + another IFRAME etc. It's a messy and slow workaround.

Is there any way to make the Python WASM code interactive with what the browser actually shows? All while not leaving the current code block.

**Comment by ryanking13** [comment]
You can run Python codes asynchronously and use `await` to pass the control back tot the browser so that the browser can do other scheduled tasks. The most easy way of it is calling `asyncio.sleep()`.
 
```
pyodide.runPythonAsync(`
  import asyncio
  await asyncio.sleep(0)
`)
```

**Comment by ghost** [comment]
Thank you for the hint using asyncio.sleep!

However, all my code is in Python, there is only one entrace point into it, so just one chance to run sync or async (and the latter never really worked).

- but when I define a DOM callback into my code (e.g. via `create_proxy()`), can I declare it as `async` to benefit from that?

**Reply by ryanking13** [reply]
> but when I define a DOM callback into my code (e.g. via create_proxy()), can I declare it as async to benefit from that?

I don't think it will probably work. The browser is single-threaded, and while the python code is running, it will continue to execute the task unless you explicitly hand over control.

Otherwise, you can use Webworker to run Python code separately from the main UI thread. But in that case, you will not be able to access the DOM (at least in a synchronous way) in Python, as webworker cannot access the DOM directly.
